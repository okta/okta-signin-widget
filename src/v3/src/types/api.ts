/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { MinimalOktaAuthIdxInterface, Tokens } from '@okta/okta-auth-js';

import {
  HooksAPI,
  RenderErrorCallback,
  RenderResult,
  RenderSuccessCallback,
  RouterEventsAPI,
} from '../../../types';
import { RenderOptions, WidgetOptions } from './widget';

// Keep a duplicate OktaSignInAPI interface because the version of auth-js differs between
// Gen3 and Gen2 so there are type mismatches in their exposed type interfaces
export interface OktaSignInAPI extends HooksAPI, RouterEventsAPI {
  // Gen3 only
  readonly options: Pick<WidgetOptions, 'brandName'>;

  // Gen3 supports only IDX
  authClient: MinimalOktaAuthIdxInterface;
  show(): void;
  hide(): void;
  remove(): void;
  showSignIn(options: RenderOptions): Promise<RenderResult>;
  showSignInToGetTokens(options: RenderOptions): Promise<Tokens>;
  showSignInAndRedirect(options: RenderOptions): Promise<void>;
  renderEl(
    options: RenderOptions,
    success?: RenderSuccessCallback,
    error?: RenderErrorCallback
  ): Promise<RenderResult>;

  getUser(): void
}
