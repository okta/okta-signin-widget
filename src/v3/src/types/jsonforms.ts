/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  IdxTransaction,
  Input,
} from '@okta/okta-auth-js';
import { FunctionComponent, h } from 'preact';

import { JsonObject } from './json';
import { FormBag } from './schema';
import { UISchemaElement } from '@jsonforms/core';

export type InputType = 'hidden' | 'password';

export enum LayoutType {
  CATEGORIZATION = 'Categorization',
  CATEGORY = 'Category',
  GROUP = 'Group',
  HORIZONTAL = 'HorizontalLayout',
  STEPPER = 'Stepper',
  VERTICAL = 'VerticalLayout',
}

export enum MessageType {
  ERROR = 'ERROR',
  WARNING = 'WARNING',
  INFO = 'INFO',
  SUCCESS = 'SUCCESS',
}
export type MessageVariant = 'error' | 'warning' | 'info' | 'success';
export const MessageTypeVariant: Record<MessageType, MessageVariant> = {
  ERROR: 'error',
  WARNING: 'warning',
  INFO: 'info',
  SUCCESS: 'success',
};

// Utility that asserts a specific key of an object non-optional
export type RequiredKeys<T, K extends keyof T> = Exclude<T, K> & Required<Pick<T, K>>;

// Utility to modify interfaces / types
export type Modify<T, R> = Omit<T, keyof R> & R;

export type TransformedResponse = {
  remediation?: FormBag[];
  cancel?: FormBag;
};

export interface WithDefaultValue<T = unknown> {
  default: T;
}

export type FieldTransformer<U = JsonObject, T = Input> = (input: T) => U | null;

export type TerminalKeyTransformer = (transaction: IdxTransaction, formBag: FormBag) => FormBag;

export type UISchemaOptions = {
  [key: string]: unknown;
};

export type Choice = {
  key: string;
  value: string;
};

export type RendererComponent<T> = {
  (props: T): h.JSX.Element;
  displayName: string;
  name?: string;
};

export type WrappedFunctionComponent<T> = (Component: FunctionComponent<T>) => RendererComponent<T>;

export interface ControlPropsWithFormValidationState {
  dirty?: boolean;
  setDirty?: (dirty: boolean) => void;
  pristine?: boolean;
  setPristine?: (pristine: boolean) => void;
  touched?: boolean;
  setTouched?: (touched: boolean) => void;
  untouched?: boolean;
  setUntouched?: (untouched: boolean) => void;
}
