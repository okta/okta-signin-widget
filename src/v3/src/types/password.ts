/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { PASSWORD_REQUIREMENTS_KEYS } from '../constants';

export type PasswordRequirementStatus = 'incomplete' | 'complete' | 'info';
export type PasswordRequirementProps = {
  status: PasswordRequirementStatus;
  label: string;
};

export interface GetAgeFromMinutes {
  unitLabel: string,
  value: number
}

export type ComplexityKeys = keyof typeof PASSWORD_REQUIREMENTS_KEYS.complexity;

export type ComplexityRequirements = {
  [key in ComplexityKeys]?: any;
};

type AgeKeys = keyof typeof PASSWORD_REQUIREMENTS_KEYS.age;
export type AgeRequirements = {
  [key in AgeKeys]?: any;
};

export interface PasswordRequirementsData {
  complexity?: ComplexityRequirements;
  age?: AgeRequirements;
}

export interface PasswordSettings {
  complexity?: {
    minLength?: number;
    minLowerCase?: number;
    minUpperCase?: number;
    minNumber?: number;
    minSymbol?: number;
    excludeUsername?: boolean;
    excludeFirstName?: boolean;
    excludeLastName?: boolean;
    excludeAttributes?: string[];
    useADComplexityRequirements?: boolean;
  };
  age?: {
    historyCount?: number;
    minAgeMinutes?: number;
    minAgeHours?: number;
    minAgeDays?: number;
  };
  daysToExpiry?: number;
}

export type PasswordValidation = {
  [key: string]: boolean;
};

export interface ListItem {
  ruleKey: string;
  label: string;
}
