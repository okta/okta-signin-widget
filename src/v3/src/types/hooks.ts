/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxAuthenticator, IdxContext, NextStep } from '@okta/okta-auth-js';

import { EventContext, HookDefinition as BaseHookDefinition, HookFunction } from '../../../types';
import type { loc } from '../util/locUtil';
import { FormBag } from './schema';
import { UserInfo } from './userInfo';

export interface DeviceEnrollment {
  name: string;
  platform: string;
  vendor?: string;
  signInUrl?: string;
  enrollmentLink?: string;
  challengeMethod?: string;
  orgName?: string;
}

export type SafeFormBag = Omit<FormBag, 'dataSchema'>;

export interface TransformHookContext extends EventContext {
  formBag: SafeFormBag;
  currentAuthenticator?: IdxAuthenticator;
  userInfo?: UserInfo;
  deviceEnrollment?: DeviceEnrollment;
  nextStep?: NextStep;
  idxContext?: IdxContext;
  loc: typeof loc,
}

export interface TransformHookFunction {
  (context: TransformHookContext): void;
}

export interface HookDefinition extends BaseHookDefinition {
  afterTransform?: TransformHookFunction[];
}

export type HookType = keyof HookDefinition;

export interface HooksOptions {
  [name: string]: HookDefinition;
}

export interface FormHooksMap extends Map<string, HookFunction[]> {
  get(k: 'afterTransform'): TransformHookFunction[];
  get(k: string): HookFunction[];
}

export type HooksMap = Map<string, FormHooksMap>;
