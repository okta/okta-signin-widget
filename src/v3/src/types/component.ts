/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { FunctionComponent } from 'preact';
import { StateUpdater } from 'preact/hooks';

import { BlurHandler } from './handlers';
import {
  ButtonElement,
  FieldElement,
  UISchemaElement,
  WidgetMessage,
} from './schema';

export type TesterFunction = (
  uiElement: UISchemaElement | FieldElement | ButtonElement
) => boolean;

export type UISchemaElementComponentWithValidationProps = {
  uischema: FieldElement;
  type?: string;
  setTouched?: StateUpdater<boolean>,
  errors?: WidgetMessage[],
  setErrors?: StateUpdater<WidgetMessage[] | undefined>,
  onValidateHandler?: (
    setErrors?: StateUpdater<WidgetMessage[] | undefined>,
    value?: string | boolean | number,
  ) => void,
  handleChange?: (value: string | number | boolean) => void;
  handleBlur?: BlurHandler;
  describedByIds?: string;
};

export type UISchemaElementComponentProps = {
  uischema: UISchemaElement;
};

export type UISchemaElementComponent<
  T = Record<string, unknown>,
> = FunctionComponent<UISchemaElementComponentProps & T>;

export type Renderer = {
  tester: TesterFunction;
  renderer: UISchemaElementComponent;
};
