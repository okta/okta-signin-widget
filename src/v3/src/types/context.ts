/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  AuthApiError,
  IdxMessage,
  IdxTransaction,
  OAuthError,
  OktaAuthIdxInterface,
} from '@okta/okta-auth-js';
import { MutableRef, StateUpdater } from 'preact/hooks';

import { FormBag, LanguageDirection, UISchemaLayoutType } from './schema';
import { WidgetProps } from './widget';

export type IWidgetContext = {
  authClient: OktaAuthIdxInterface;
  widgetProps: WidgetProps;
  message: IdxMessage | undefined;
  setMessage: StateUpdater<IdxMessage | undefined>;
  onSuccessCallback?: WidgetProps['globalSuccessFn'];
  onErrorCallback?: WidgetProps['globalErrorFn'];
  idxTransaction: IdxTransaction | undefined;
  setResponseError: StateUpdater<AuthApiError | OAuthError | null>;
  setIdxTransaction: StateUpdater<IdxTransaction | undefined>;
  setIsClientTransaction: StateUpdater<boolean>;
  stepToRender: string | undefined;
  setStepToRender: StateUpdater<string | undefined>;
  data: FormBag['data'];
  setData: StateUpdater<FormBag['data']>;
  dataSchemaRef: MutableRef<FormBag['dataSchema'] | undefined>;
  loading: boolean;
  setLoading: StateUpdater<boolean>;
  setWidgetRendered: StateUpdater<boolean>;
  loginHint?: string | null;
  setloginHint: StateUpdater<string | null>;
  languageCode: string;
  languageDirection: LanguageDirection;
};

// Stepper context
export type IStepperContext = {
  stepIndex: number;
  setStepIndex: StateUpdater<number | undefined>;
};

// Layout context
export type ILayoutContext = {
  layoutDirection: UISchemaLayoutType.VERTICAL | UISchemaLayoutType.HORIZONTAL,
};
