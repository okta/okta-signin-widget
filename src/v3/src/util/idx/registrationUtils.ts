/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  APIError, FieldError, IdxMessage, Input,
} from '@okta/okta-auth-js';
import { StateUpdater } from 'preact/hooks';

import { RegistrationElementSchema, WidgetMessage } from '../../types';
import { flattenInputs } from '../flattenInputs';
import { loc } from '../locUtil';
import { resetMessagesToInputs } from '../resetMessagesToInputs';

type RegistrationFieldError = FieldError & { property: string };

export const convertIdxInputsToRegistrationSchema = (
  inputs: Input[],
): RegistrationElementSchema[] => (
  inputs.reduce((acc: Input[], input: Input) => {
    const flattenedInputs = flattenInputs(input);
    return [...acc, ...flattenedInputs];
  }, []).map((flattenedInput: Input) => ({
    ...flattenedInput,
    'label-top': true,
    'data-se': flattenedInput.name,
    wide: true,
  } as RegistrationElementSchema))
);

export const convertRegistrationSchemaToIdxInputs = (
  schema: RegistrationElementSchema[],
  idxInputs: Input[],
): void => {
  schema.forEach((schemaEle: RegistrationElementSchema) => {
    if (schemaEle.name?.includes('.')) {
      // nested object (only go one level deep)
      const [schemaGroupName, schemaGrpFieldName] = schemaEle.name.split('.');
      const inputGroupIndex = idxInputs.findIndex(({ name }) => name === schemaGroupName);
      if (inputGroupIndex !== -1 && Array.isArray(idxInputs[inputGroupIndex]?.value)) {
        // Existing input
        const inputGroupFieldIndex = (
          idxInputs[inputGroupIndex]?.value as Input[]
        )?.findIndex(({ name }) => name === schemaGrpFieldName);
        if (inputGroupFieldIndex !== -1 && Array.isArray(idxInputs[inputGroupIndex].value)) {
          // Update existing field
          const srcObj = (idxInputs[inputGroupIndex].value as Input[])[inputGroupFieldIndex];
          Object.assign(srcObj, { ...schemaEle, name: schemaGrpFieldName });
        } else {
          // Add new field
          const schemaInputField = { ...schemaEle, name: schemaGrpFieldName };
          (idxInputs[inputGroupIndex].value as Record<string, unknown>[]).push(schemaInputField);
        }
      } else {
        // New top level input
        idxInputs.push({
          name: schemaGroupName,
          type: 'object',
          required: schemaEle.required,
          // @ts-expect-error Registration Schema Element has extra fields that are not in Input
          value: { ...schemaEle, name: schemaGrpFieldName },
        });
      }
    } else {
      // not nested
      const existFieldIdx = idxInputs.findIndex(({ name }) => name === schemaEle.name);
      if (existFieldIdx !== -1) {
        // Existing Field (modify)
        Object.assign(idxInputs[existFieldIdx], schemaEle);
      } else {
        // New Field (create)
        // @ts-expect-error Registration Schema Element has extra fields that are not in Input
        idxInputs.push(schemaEle);
      }
    }
  });
};

export const triggerRegistrationErrorMessages = (
  error: APIError,
  inputs: Input[],
  setMessage: StateUpdater<IdxMessage | undefined>,
): void => {
  if (Array.isArray(error.errorCauses)) {
    // field level error messages
    const messagesByField: Record<string, WidgetMessage[]> = error.errorCauses
      .reduce((acc, err) => {
        const { errorSummary, property } = err as RegistrationFieldError;
        return {
          ...acc,
          [property]: [{ class: 'ERROR', message: errorSummary }],
        };
      }, {});
    resetMessagesToInputs(inputs, messagesByField);
  }

  setMessage({
    class: 'ERROR',
    i18n: { key: '' },
    message: error.errorSummary || loc('oform.errorbanner.title', 'login'),
  });
};
