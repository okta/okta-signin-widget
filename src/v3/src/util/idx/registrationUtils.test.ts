/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { APIError, Input } from '@okta/okta-auth-js';

import { RegistrationElementSchema } from '../../types';
import { convertIdxInputsToRegistrationSchema, convertRegistrationSchemaToIdxInputs, triggerRegistrationErrorMessages } from './registrationUtils';

describe('idx/registrationUtils', () => {
  describe('convertIdxInputsToRegistrationSchema', () => {
    it('should not perform conversion of Idx Inputs into Registration schema elements when input array is empty', () => {
      expect(convertIdxInputsToRegistrationSchema([])).toEqual([]);
    });

    it('should convert array of Idx Input objects into an array of Registration schema objects', () => {
      const inputs: Input[] = [
        {
          name: 'userProfile',
          value: [
            { name: 'firstName', label: 'First name', required: true },
            { name: 'lastName', label: 'Last name', required: true },
            { name: 'email', label: 'Email', required: true },
          ],
        },
        {
          name: 'credentials',
          value: [{
            name: 'passcode',
            label: 'Password',
            required: false,
            secret: true,
          }],
        },
      ];
      const schemaElements: RegistrationElementSchema[] = convertIdxInputsToRegistrationSchema(
        inputs,
      );

      expect(schemaElements.length).toBe(4);
      expect(schemaElements).toMatchSnapshot();
      expect(schemaElements[0].name).toBe('userProfile.firstName');
      expect(schemaElements[1].name).toBe('userProfile.lastName');
      expect(schemaElements[2].name).toBe('userProfile.email');
      expect(schemaElements[3].name).toBe('credentials.passcode');
    });
  });

  describe('convertRegistrationSchemaToIdxInputs', () => {
    it('should update Idx Inputs array with modified registration schema elements when adding new field and updating existing', () => {
      const inputs: Input[] = [
        {
          name: 'userProfile',
          value: [
            { name: 'firstName', label: 'First name', required: true },
            { name: 'lastName', label: 'Last name', required: true },
            { name: 'email', label: 'Email', required: true },
          ],
        },
        {
          name: 'credentials',
          value: [{
            name: 'passcode',
            label: 'Password',
            required: false,
            secret: true,
          }],
        },
      ];
      const schema: RegistrationElementSchema[] = [
        {
          name: 'userProfile.firstName',
          label: 'Given name',
          required: false,
          'label-top': true,
          'data-se': 'userProfile.firstName',
          wide: true,
        },
        {
          name: 'userProfile.lastName',
          label: 'Family name',
          required: false,
          'label-top': true,
          'data-se': 'userProfile.lastName',
          wide: true,
        },
        {
          name: 'userProfile.email',
          label: 'Company login',
          required: false,
          'label-top': true,
          'data-se': 'userProfile.email',
          wide: true,
        },
        {
          name: 'userProfile.favoriteCar',
          label: 'Favorite car',
          required: true,
          'label-top': true,
          'data-se': 'userProfile.favoriteCar',
          wide: true,
        },
        {
          name: 'credentials.passcode',
          label: 'Password',
          required: true,
          'label-top': true,
          'data-se': 'credentials.passcode',
          wide: true,
        },
      ];

      convertRegistrationSchemaToIdxInputs(schema, inputs);

      expect(inputs.length).toBe(2);
      expect((inputs[0].value as Input[]).length).toBe(4);
      expect((inputs[1].value as Input[]).length).toBe(1);
      expect(inputs).toMatchSnapshot();
      expect((inputs[0].value as Input[])[0].name).toBe('firstName');
      expect((inputs[0].value as Input[])[1].name).toBe('lastName');
      expect((inputs[0].value as Input[])[2].name).toBe('email');
      expect((inputs[0].value as Input[])[3].name).toBe('favoriteCar');
    });
  });

  describe('triggerRegistrationErrorMessages', () => {
    it('should trigger custom global registration error message when triggering error with errorSummary', () => {
      const mockSetMessageFn = jest.fn();
      const inputs: Input[] = [
        {
          name: 'userProfile',
          value: [
            { name: 'firstName', label: 'First name', required: true },
            { name: 'lastName', label: 'Last name', required: true },
            { name: 'email', label: 'Email', required: true },
          ],
        },
        {
          name: 'credentials',
          value: [{
            name: 'passcode',
            label: 'Password',
            required: false,
            secret: true,
          }],
        },
      ];
      triggerRegistrationErrorMessages({ errorSummary: 'This is a custom global error message' }, inputs, mockSetMessageFn);

      expect(mockSetMessageFn).toHaveBeenCalledWith({
        class: 'ERROR',
        i18n: { key: '' },
        message: 'This is a custom global error message',
      });
    });

    it('should trigger field level registration error messages when triggering error with errorCauses', () => {
      const mockSetMessageFn = jest.fn();
      const inputs: Input[] = [
        {
          name: 'userProfile',
          value: [
            { name: 'firstName', label: 'First name', required: true },
            { name: 'lastName', label: 'Last name', required: true },
            { name: 'email', label: 'Email', required: true },
          ],
        },
        {
          name: 'credentials',
          value: [{
            name: 'passcode',
            label: 'Password',
            required: false,
            secret: true,
          }],
        },
      ];
      triggerRegistrationErrorMessages(
        // @ts-expect-error property is expected in SIW but not defined in auth-js type
        { errorCauses: [{ property: 'userProfile.lastName', errorSummary: 'Custom field level error' }] },
        inputs,
        mockSetMessageFn,
      );

      // @ts-expect-error OKTA-539834 - messages missing from type
      expect((inputs[0].value as Input[])[1].messages?.value).toEqual([{
        class: 'ERROR', message: 'Custom field level error',
      }]);
      expect(mockSetMessageFn).toHaveBeenCalledWith({
        class: 'ERROR',
        i18n: { key: '' },
        message: 'oform.errorbanner.title',
      });
    });

    it('should trigger default global registration error message when triggering error without errorSummary', () => {
      const mockSetMessageFn = jest.fn();
      const inputs: Input[] = [
        {
          name: 'userProfile',
          value: [
            { name: 'firstName', label: 'First name', required: true },
            { name: 'lastName', label: 'Last name', required: true },
            { name: 'email', label: 'Email', required: true },
          ],
        },
        {
          name: 'credentials',
          value: [{
            name: 'passcode',
            label: 'Password',
            required: false,
            secret: true,
          }],
        },
      ];
      triggerRegistrationErrorMessages({ errorCode: 'E0047' } as APIError, inputs, mockSetMessageFn);

      expect(mockSetMessageFn).toHaveBeenCalledWith({
        class: 'ERROR',
        i18n: { key: '' },
        message: 'oform.errorbanner.title',
      });
    });
  });
});
