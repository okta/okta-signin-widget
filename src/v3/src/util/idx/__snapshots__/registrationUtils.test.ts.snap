// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idx/registrationUtils convertIdxInputsToRegistrationSchema should convert array of Idx Input objects into an array of Registration schema objects 1`] = `
Array [
  Object {
    "data-se": "userProfile.firstName",
    "label": "First name",
    "label-top": true,
    "name": "userProfile.firstName",
    "required": true,
    "wide": true,
  },
  Object {
    "data-se": "userProfile.lastName",
    "label": "Last name",
    "label-top": true,
    "name": "userProfile.lastName",
    "required": true,
    "wide": true,
  },
  Object {
    "data-se": "userProfile.email",
    "label": "Email",
    "label-top": true,
    "name": "userProfile.email",
    "required": true,
    "wide": true,
  },
  Object {
    "data-se": "credentials.passcode",
    "label": "Password",
    "label-top": true,
    "name": "credentials.passcode",
    "required": false,
    "secret": true,
    "wide": true,
  },
]
`;

exports[`idx/registrationUtils convertRegistrationSchemaToIdxInputs should update Idx Inputs array with modified registration schema elements when adding new field and updating existing 1`] = `
Array [
  Object {
    "name": "userProfile",
    "value": Array [
      Object {
        "data-se": "userProfile.firstName",
        "label": "Given name",
        "label-top": true,
        "name": "firstName",
        "required": false,
        "wide": true,
      },
      Object {
        "data-se": "userProfile.lastName",
        "label": "Family name",
        "label-top": true,
        "name": "lastName",
        "required": false,
        "wide": true,
      },
      Object {
        "data-se": "userProfile.email",
        "label": "Company login",
        "label-top": true,
        "name": "email",
        "required": false,
        "wide": true,
      },
      Object {
        "data-se": "userProfile.favoriteCar",
        "label": "Favorite car",
        "label-top": true,
        "name": "favoriteCar",
        "required": true,
        "wide": true,
      },
    ],
  },
  Object {
    "name": "credentials",
    "value": Array [
      Object {
        "data-se": "credentials.passcode",
        "label": "Password",
        "label-top": true,
        "name": "passcode",
        "required": true,
        "secret": true,
        "wide": true,
      },
    ],
  },
]
`;
